File: Delphi-source-code/MyInventory/MyInventoryStock/src/uInventoryStockBL.pas
unit uInventoryStockBL;

interface

uses
  System.SysUtils;

type
  TStockItem = record
    ItemId: Integer;
    ItemName: string;
    Category: string;
    UnitMeasure: string;
    CurrentQty: Double;
    MinStock: Double;
    UnitCost: Currency;
    NeedsReorder: Boolean;
  end;

  TStockMovement = record
    MovementId: Integer;
    ItemName: string;
    MovementType: string;
    Quantity: Double;
    MovementDate: TDateTime;
    Reference: string;
  end;

  TStockValuation = record
    Category: string;
    ItemCount: Integer;
    TotalQuantity: Double;
    TotalValue: Currency;
  end;

  TInventoryStockBL = class
  public
    function GetStockItems: TArray<TStockItem>;
    function GetStockMovements: TArray<TStockMovement>;
    function GetStockValuation: TArray<TStockValuation>;
  end;

const
  SQL_GET_STOCK_ITEMS =
    'SELECT si.ItemId, si.ItemName, si.Category, si.UnitMeasure, ' +
    '       si.CurrentQty, si.MinStock, si.UnitCost, ' +
    '       CASE WHEN si.CurrentQty <= si.MinStock THEN 1 ELSE 0 END AS NeedsReorder ' +
    'FROM StockItems si ' +
    'ORDER BY si.Category, si.ItemName';

  SQL_GET_STOCK_MOVEMENTS =
    'SELECT sm.MovementId, si.ItemName, sm.MovementType, sm.Quantity, ' +
    '       sm.MovementDate, sm.Reference ' +
    'FROM StockMovements sm ' +
    'INNER JOIN StockItems si ON si.ItemId = sm.ItemId ' +
    'ORDER BY sm.MovementDate DESC';

  SQL_GET_STOCK_VALUATION =
    'SELECT si.Category, COUNT(*) AS ItemCount, ' +
    '       SUM(si.CurrentQty) AS TotalQuantity, ' +
    '       SUM(si.CurrentQty * si.UnitCost) AS TotalValue ' +
    'FROM StockItems si ' +
    'GROUP BY si.Category ' +
    'ORDER BY TotalValue DESC';

  SQL_UPDATE_STOCK_QTY =
    'UPDATE StockItems SET CurrentQty = :NewQty, ' +
    '       LastUpdated = GETDATE() ' +
    'WHERE ItemId = :ItemId';

implementation

{ TInventoryStockBL }

function TInventoryStockBL.GetStockItems: TArray<TStockItem>;

  procedure AddItem(var Arr: TArray<TStockItem>; AId: Integer;
    const AName, ACategory, AUnit: string; AQty, AMin: Double; ACost: Currency);
  var
    Idx: Integer;
  begin
    Idx := Length(Arr);
    SetLength(Arr, Idx + 1);
    Arr[Idx].ItemId := AId;
    Arr[Idx].ItemName := AName;
    Arr[Idx].Category := ACategory;
    Arr[Idx].UnitMeasure := AUnit;
    Arr[Idx].CurrentQty := AQty;
    Arr[Idx].MinStock := AMin;
    Arr[Idx].UnitCost := ACost;
    Arr[Idx].NeedsReorder := AQty <= AMin;
  end;

begin
  Result := nil;
  AddItem(Result, 1,  'Olive Oil',     'Oils',       'liters', 12, 5,  18.50);
  AddItem(Result, 2,  'Flour',         'Dry Goods',  'kg',     45, 20,  2.80);
  AddItem(Result, 3,  'Tomatoes',      'Produce',    'kg',     30, 10,  3.20);
  AddItem(Result, 4,  'Mozzarella',    'Dairy',      'kg',     15, 5,  12.00);
  AddItem(Result, 5,  'Salmon Fillet', 'Seafood',    'kg',      8, 3,  22.00);
  AddItem(Result, 6,  'Chicken Breast','Meat',       'kg',     20, 8,   9.50);
  AddItem(Result, 7,  'Garlic',        'Produce',    'kg',      5, 2,   6.00);
  AddItem(Result, 8,  'Basil',         'Herbs',      'kg',      3, 1,  15.00);
  AddItem(Result, 9,  'Parmesan',      'Dairy',      'kg',     10, 3,  28.00);
  AddItem(Result, 10, 'Rice',          'Dry Goods',  'kg',     25, 10,  4.50);
  AddItem(Result, 11, 'Butter',        'Dairy',      'kg',      8, 3,   7.50);
  AddItem(Result, 12, 'Lemons',        'Produce',    'kg',      6, 2,   4.00);
  AddItem(Result, 13, 'Wine Vinegar',  'Condiments', 'liters',  4, 2,   8.00);
  AddItem(Result, 14, 'Black Pepper',  'Spices',     'kg',      2, 1,  35.00);
  AddItem(Result, 15, 'Heavy Cream',   'Dairy',      'liters', 10, 4,   5.50);
end;

function TInventoryStockBL.GetStockMovements: TArray<TStockMovement>;

  procedure AddMovement(var Arr: TArray<TStockMovement>; AId: Integer;
    const AName, AType: string; AQty: Double; ADate: TDateTime;
    const ARef: string);
  var
    Idx: Integer;
  begin
    Idx := Length(Arr);
    SetLength(Arr, Idx + 1);
    Arr[Idx].MovementId := AId;
    Arr[Idx].ItemName := AName;
    Arr[Idx].MovementType := AType;
    Arr[Idx].Quantity := AQty;
    Arr[Idx].MovementDate := ADate;
    Arr[Idx].Reference := ARef;
  end;

begin
  Result := nil;
  AddMovement(Result, 1,  'Olive Oil',      'IN',  24,  EncodeDate(2026, 7, 28), 'PO-2026-045');
  AddMovement(Result, 2,  'Flour',          'IN',  50,  EncodeDate(2026, 7, 27), 'PO-2026-044');
  AddMovement(Result, 3,  'Tomatoes',       'OUT', 8,   EncodeDate(2026, 8, 1),  'Order #1023');
  AddMovement(Result, 4,  'Mozzarella',     'OUT', 5,   EncodeDate(2026, 8, 1),  'Order #1024');
  AddMovement(Result, 5,  'Salmon Fillet',  'IN',  12,  EncodeDate(2026, 7, 30), 'PO-2026-046');
  AddMovement(Result, 6,  'Chicken Breast', 'OUT', 10,  EncodeDate(2026, 8, 2),  'Order #1025');
  AddMovement(Result, 7,  'Basil',          'OUT', 1.5, EncodeDate(2026, 8, 2),  'Order #1026');
  AddMovement(Result, 8,  'Parmesan',       'IN',  15,  EncodeDate(2026, 7, 25), 'PO-2026-042');
  AddMovement(Result, 9,  'Rice',           'OUT', 5,   EncodeDate(2026, 8, 3),  'Order #1027');
  AddMovement(Result, 10, 'Heavy Cream',    'IN',  20,  EncodeDate(2026, 7, 29), 'PO-2026-043');
end;

function TInventoryStockBL.GetStockValuation: TArray<TStockValuation>;

  procedure AddValuation(var Arr: TArray<TStockValuation>;
    const ACategory: string; ACount: Integer; AQty: Double; AValue: Currency);
  var
    Idx: Integer;
  begin
    Idx := Length(Arr);
    SetLength(Arr, Idx + 1);
    Arr[Idx].Category := ACategory;
    Arr[Idx].ItemCount := ACount;
    Arr[Idx].TotalQuantity := AQty;
    Arr[Idx].TotalValue := AValue;
  end;

begin
  Result := nil;
  AddValuation(Result, 'Dairy',      4, 43, 601.00);
  AddValuation(Result, 'Produce',    3, 41, 228.00);
  AddValuation(Result, 'Oils',       1, 12, 222.00);
  AddValuation(Result, 'Meat',       1, 20, 190.00);
  AddValuation(Result, 'Seafood',    1,  8, 176.00);
  AddValuation(Result, 'Dry Goods',  2, 70, 238.50);
  AddValuation(Result, 'Spices',     1,  2,  70.00);
  AddValuation(Result, 'Herbs',      1,  3,  45.00);
  AddValuation(Result, 'Condiments', 1,  4,  32.00);
end;

end.

File: MyInventoryStock.Core/Application/Services/MyInventoryStock/uInventoryStockBL.cs
using Microsoft.Extensions.Logging;
using MyInventoryStock.Core.Application.DTOs.Core;

namespace MyInventoryStock.Core.Application.Services.MyInventoryStock;

public class uInventoryStockBL : IuInventoryStockBL
{
    private readonly ILogger<uInventoryStockBL> _logger;

    public uInventoryStockBL(ILogger<uInventoryStockBL> logger)
    {
        _logger = logger;
    }

    /// <summary>
    /// Synchronous version ported from TInventoryStockBL.GetStockItems.
    /// Returns the full stock item catalog with reorder flags.
    /// </summary>
    public List<TStockItem> GetStockItems()
    {
        _logger.LogDebug("GetStockItems called");
        return BuildStockItemList();
    }

    /// <summary>
    /// Synchronous version ported from TInventoryStockBL.GetStockMovements.
    /// Returns recent stock movements (receipts and issues).
    /// </summary>
    public List<TStockMovement> GetStockMovements()
    {
        _logger.LogDebug("GetStockMovements called");
        return BuildStockMovementList();
    }

    /// <summary>
    /// Synchronous version ported from TInventoryStockBL.GetStockValuation.
    /// Returns stock valuation aggregated by category.
    /// </summary>
    public List<TStockValuation> GetStockValuation()
    {
        _logger.LogDebug("GetStockValuation called");
        return BuildStockValuationList();
    }

    /// <summary>
    /// Async version ported from TInventoryStockBL.GetStockItems.
    /// Returns the full stock item catalog with reorder flags.
    /// Legacy used hardcoded seed data; in production this would query StockItems table
    /// using SQL_GET_STOCK_ITEMS.
    /// </summary>
    public async Task<TStockItem[]> GetStockItemsAsync(CancellationToken ct = default)
    {
        _logger.LogDebug("GetStockItemsAsync called");
        await Task.CompletedTask;
        return BuildStockItemList().ToArray();
    }

    /// <summary>
    /// Async version ported from TInventoryStockBL.GetStockMovements.
    /// Returns recent stock movements (receipts and issues).
    /// Legacy used hardcoded seed data; in production this would query StockMovements
    /// joined with StockItems using SQL_GET_STOCK_MOVEMENTS.
    /// </summary>
    public async Task<TStockMovement[]> GetStockMovementsAsync(CancellationToken ct = default)
    {
        _logger.LogDebug("GetStockMovementsAsync called");
        await Task.CompletedTask;
        return BuildStockMovementList().ToArray();
    }

    /// <summary>
    /// Async version ported from TInventoryStockBL.GetStockValuation.
    /// Returns stock valuation aggregated by category.
    /// Legacy used hardcoded seed data; in production this would query StockItems
    /// with GROUP BY using SQL_GET_STOCK_VALUATION.
    /// </summary>
    public async Task<TStockValuation[]> GetStockValuationAsync(CancellationToken ct = default)
    {
        _logger.LogDebug("GetStockValuationAsync called");
        await Task.CompletedTask;
        return BuildStockValuationList().ToArray();
    }

    private static List<TStockItem> BuildStockItemList()
    {
        return new List<TStockItem>
        {
            BuildStockItem(1,  "Olive Oil",      "Oils",       "liters", 12m,  5m,  18.50m),
            BuildStockItem(2,  "Flour",           "Dry Goods",  "kg",     45m, 20m,   2.80m),
            BuildStockItem(3,  "Tomatoes",        "Produce",    "kg",     30m, 10m,   3.20m),
            BuildStockItem(4,  "Mozzarella",      "Dairy",      "kg",     15m,  5m,  12.00m),
            BuildStockItem(5,  "Salmon Fillet",   "Seafood",    "kg",      8m,  3m,  22.00m),
            BuildStockItem(6,  "Chicken Breast",  "Meat",       "kg",     20m,  8m,   9.50m),
            BuildStockItem(7,  "Garlic",          "Produce",    "kg",      5m,  2m,   6.00m),
            BuildStockItem(8,  "Basil",           "Herbs",      "kg",      3m,  1m,  15.00m),
            BuildStockItem(9,  "Parmesan",        "Dairy",      "kg",     10m,  3m,  28.00m),
            BuildStockItem(10, "Rice",            "Dry Goods",  "kg",     25m, 10m,   4.50m),
            BuildStockItem(11, "Butter",          "Dairy",      "kg",      8m,  3m,   7.50m),
            BuildStockItem(12, "Lemons",          "Produce",    "kg",      6m,  2m,   4.00m),
            BuildStockItem(13, "Wine Vinegar",    "Condiments", "liters",  4m,  2m,   8.00m),
            BuildStockItem(14, "Black Pepper",    "Spices",     "kg",      2m,  1m,  35.00m),
            BuildStockItem(15, "Heavy Cream",     "Dairy",      "liters", 10m,  4m,   5.50m),
        };
    }

    private static List<TStockMovement> BuildStockMovementList()
    {
        return new List<TStockMovement>
        {
            BuildStockMovement(1,  "Olive Oil",      "IN",  24.0m,  new DateTime(2026, 7, 28, 0, 0, 0, DateTimeKind.Utc), "PO-2026-045"),
            BuildStockMovement(2,  "Flour",           "IN",  50.0m,  new DateTime(2026, 7, 27, 0, 0, 0, DateTimeKind.Utc), "PO-2026-044"),
            BuildStockMovement(3,  "Tomatoes",        "OUT",  8.0m,  new DateTime(2026, 8, 1, 0, 0, 0, DateTimeKind.Utc),  "Order #1023"),
            BuildStockMovement(4,  "Mozzarella",      "OUT",  5.0m,  new DateTime(2026, 8, 1, 0, 0, 0, DateTimeKind.Utc),  "Order #1024"),
            BuildStockMovement(5,  "Salmon Fillet",   "IN",  12.0m,  new DateTime(2026, 7, 30, 0, 0, 0, DateTimeKind.Utc), "PO-2026-046"),
            BuildStockMovement(6,  "Chicken Breast",  "OUT", 10.0m,  new DateTime(2026, 8, 2, 0, 0, 0, DateTimeKind.Utc),  "Order #1025"),
            BuildStockMovement(7,  "Basil",           "OUT",  1.5m,  new DateTime(2026, 8, 2, 0, 0, 0, DateTimeKind.Utc),  "Order #1026"),
            BuildStockMovement(8,  "Parmesan",        "IN",  15.0m,  new DateTime(2026, 7, 25, 0, 0, 0, DateTimeKind.Utc), "PO-2026-042"),
            BuildStockMovement(9,  "Rice",            "OUT",  5.0m,  new DateTime(2026, 8, 3, 0, 0, 0, DateTimeKind.Utc),  "Order #1027"),
            BuildStockMovement(10, "Heavy Cream",     "IN",  20.0m,  new DateTime(2026, 7, 29, 0, 0, 0, DateTimeKind.Utc), "PO-2026-043"),
        };
    }

    private static List<TStockValuation> BuildStockValuationList()
    {
        return new List<TStockValuation>
        {
            BuildStockValuation("Dairy",      4, 43.0m, 601.00m),
            BuildStockValuation("Produce",    3, 41.0m, 228.00m),
            BuildStockValuation("Oils",       1, 12.0m, 222.00m),
            BuildStockValuation("Meat",       1, 20.0m, 190.00m),
            BuildStockValuation("Seafood",    1,  8.0m, 176.00m),
            BuildStockValuation("Dry Goods",  2, 70.0m, 238.50m),
            BuildStockValuation("Spices",     1,  2.0m,  70.00m),
            BuildStockValuation("Herbs",      1,  3.0m,  45.00m),
            BuildStockValuation("Condiments", 1,  4.0m,  32.00m),
        };
    }

    /// <summary>
    /// Mirrors the legacy AddItem nested procedure.
    /// NeedsReorder is true when CurrentQty &lt;= MinStock (legacy: AQty &lt;= AMin).
    /// </summary>
    private static TStockItem BuildStockItem(
        int itemId,
        string itemName,
        string category,
        string unitMeasure,
        decimal currentQty,
        decimal minStock,
        decimal unitCost)
    {
        return new TStockItem
        {
            ItemId = itemId,
            ItemName = itemName,
            Category = category,
            UnitMeasure = unitMeasure,
            CurrentQty = (double)(currentQty),
            MinStock = (double)(minStock),
            UnitCost = unitCost,
            NeedsReorder = currentQty <= minStock
        };
    }

    /// <summary>
    /// Mirrors the legacy AddMovement nested procedure.
    /// </summary>
    private static TStockMovement BuildStockMovement(
        int movementId,
        string itemName,
        string movementType,
        decimal quantity,
        DateTime movementDate,
        string reference)
    {
        return new TStockMovement
        {
            MovementId = movementId,
            ItemName = itemName,
            MovementType = movementType,
            Quantity = (double)(quantity),
            MovementDate = movementDate,
            Reference = reference
        };
    }

    /// <summary>
    /// Mirrors the legacy AddValuation nested procedure.
    /// </summary>
    private static TStockValuation BuildStockValuation(
        string category,
        int itemCount,
        decimal totalQuantity,
        decimal totalValue)
    {
        return new TStockValuation
        {
            Category = category,
            ItemCount = itemCount,
            TotalQuantity = (double)(totalQuantity),
            TotalValue = totalValue
        };
    }
}
